front/pages/api/login.ts

```diff
@@ -187,79 +137,158 @@ async function handler(
       lastName,
     });
 
-    autoJoinWorkspaceWithDomain = await findWorkspaceWithWhitelistedDomain(
-      session
+    return { isCreated: true, user };
+  }
+}
+
+// `membershipInvite` flow: we know we can add the user to the associated `workspaceId` as
+// all the checks (decoding the JWT) have been run before. Simply create the membership if
+// it does not already exist and mark the invitation as consumed.
+async function handleMembershipInvite(
+  user: User,
+  membershipInvite: MembershipInvitation
+) {
+  if (membershipInvite.inviteEmail !== user.email) {
```

Comment:
Main change, we now strongly assert that the user's email is the same as the invitation.